ArduPlane/stall_detection.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2020  Kraus-Hamdani Aerospace. All rights reserved.
+ *
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "Plane.h"
+
+/*
+  stall detection algorithm and anti-stall management
+ */
+
+/*
+  Do we think we are stalled?
+  Probabilistic method where a bool is low-passed and considered a probability.
+*/
+void Plane::stall_detection_update(void)
+{
+    stall_state.raw_algorithm_output = false;
+
+    if (!arming.is_armed() ||
+        auto_state.last_flying_ms == 0 ||
+        !is_flying() ||
+        crash_state.is_crashed)
+    {
+        stall_state.stall_clear();
+        stall_state.count = 0;
+        stall_detection_log();
+        return;
+    }
+
+    if (stall_state.inhibit_ms != 0) {
+        // if someone told us to inihibt, freeze all stall processing for a brief time.
+        // This is useful when changing waypoints where we will likely perform a
+        // hard turn and roll/pitch error spikes for a moment
+        if (AP_HAL::millis() - stall_state.inhibit_ms > 2000) {
+            stall_state.inhibit_ms = 0;
+        }
+        stall_detection_log();
+        return;
+    }
+
+    const bool is_stalled_initial = stall_state.is_stalled();
+
+    stall_state.raw_algorithm_output = stall_detection_algorithm(true);
+
+    // LPF confidence
+    stall_state.confidence = ((float)stall_state.raw_algorithm_output * stall_state.coef) + (stall_state.confidence * (1.0f - stall_state.coef));
+
+    if ((is_stalled_initial != stall_state.is_stalled()) && stall_state.is_stalled()) {
+        // we just stalled
+
+        gcs().send_text_rate_limited(MAV_SEVERITY_WARNING, 1000, stall_state.detect_notify_gcs_last_ms, "STALL DETECTED!");
```

Comment:
This can probably be removed now. This is useful when experimenting with the algorithms and it auto unstall and restall very quickly. The STALL_DUR_MIN feature solved this.

Reply:
but until then.. there remains a possibility that confidence can ride the > 0.5 line and go back and forth to spam the GCS